producers/kubernetes/ingress.go

```diff
@@ -67,65 +67,80 @@ func (a *kubernetesIngressProducer) Endpoints() ([]*pkg.Endpoint, error) {
 	return endpoints, nil
 }
 
-func (a *kubernetesIngressProducer) StartWatch() error {
-	w, err := a.client.Ingresses(api.NamespaceAll).Watch(api.ListOptions{})
-	if err != nil {
-		return fmt.Errorf("Unable to watch list of ingress: %v", err)
-	}
-
-	for event := range w.ResultChan() {
-		if event.Type == watch.Error {
-			// TODO: consider allowing the service continue running and just log this error
-			return fmt.Errorf("Event listener received an error, terminating: %v", event)
-		}
-
-		if event.Type != watch.Added && event.Type != watch.Modified {
-			continue
-		}
+func (a *kubernetesIngressProducer) Monitor(results chan *pkg.Endpoint, errChan chan error, done chan struct{}, wg *sync.WaitGroup) {
+	wg.Add(1)
+	defer wg.Done()
 
-		ing, ok := event.Object.(*extensions.Ingress)
-		if !ok {
-			// If the object wasn't a Service we can safely ignore it
-			log.Printf("Cannot cast object to ingress: %v", ing)
-			continue
-		}
-
-		log.Printf("%s: %s/%s", event.Type, ing.Namespace, ing.Name)
-
-		if err := validateIngress(*ing); err != nil {
-			log.Warnln(err)
-			continue
-		}
-
-		eps, err := a.convertIngressToEndpoint(*ing)
+	for {
+		w, err := a.client.Ingresses(api.NamespaceAll).Watch(api.ListOptions{})
 		if err != nil {
-			// TODO: consider letting the service continue running and just log this error
-			return err
+			errChan <- fmt.Errorf("[Ingress] Unable to watch list of ingress: %v", err)
+			select {
+			case <-done:
+				log.Info("[Ingress] Exited monitoring loop.")
+				return
+			default:
+				time.Sleep(5 * time.Second)
+				continue
+			}
 		}
 
-		for _, ep := range eps {
-			a.channel <- ep
+		for {
+			select {
+			case event := <-w.ResultChan():
+				if event.Type == watch.Error {
+					// TODO: consider allowing the service continue running and just log this error
+					errChan <- fmt.Errorf("[Ingress] Event listener received an error, terminating: %v", event)
```

Comment:
thanks. the `continue` is also wrong as it continues inside the `select` not the `for` loop. I used a `goto` statement to restart from the outer-most `for` loop. yes, `goto`..., but it works very well in this case.